en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/01_Shader_modules.adoc

```diff
@@ -2,248 +2,362 @@
-If you happen to write non-trivial shaders with a GPU from one of these vendors, then you'd risk other vendor's drivers rejecting your code due to syntax errors, or worse, your shader running differently because of compiler bugs.
-With a straightforward bytecode format like SPIR-V that will hopefully be avoided.
-
-However, that does not mean that we need to write this bytecode by hand.
-Khronos has released their own vendor-independent compiler that compiles GLSL to SPIR-V.
-This compiler is designed to verify that your shader code is fully standards compliant and produces one SPIR-V binary that you can ship with your program.
-You can also include this compiler as a library to produce SPIR-V at runtime, but we won't be doing that in this tutorial.
-Although we can use this compiler directly via `glslangValidator.exe`, we will be using `glslc.exe` by Google instead.
-The advantage of `glslc` is that it uses the same parameter format as well-known compilers like GCC and Clang and includes some extra functionality like _includes_.
-Both of them are already included in the Vulkan SDK, so you don't need to download anything extra.
-
-GLSL is a shading language with a C-style syntax.
-Programs written in it have a `main` function that is invoked for every object.
-Instead of using parameters for input and a return value as output, GLSL uses global variables to handle input and output.
-The language includes many features to aid in graphics programming, like built-in vector and matrix primitives.
-Functions for operations like cross products, matrix-vector products and reflections around a vector are included.
-The vector type is called `vec` with a number indicating the amount of elements.
-For example, a 3D position would be stored in a `vec3`.
-It is possible to access single components through members like `.x`, but it's also possible to create a new vector from multiple components at the same time.
-For example, the expression `vec3(1.0, 2.0, 3.0).xy` would result in `vec2`.
-The constructors of vectors can also take combinations of vector objects and scalar values.
-For example, a `vec3` can be constructed with `vec3(vec2(1.0, 2.0), 3.0)`.
-
-As the previous chapter mentioned, we need to write a vertex shader and a fragment shader to get a triangle on the screen.
-The next two sections will cover the GLSL code of each of those and after that I'll show you how to produce two SPIR-V binaries and load them into the program.
+Unlike earlier APIs, shader code in Vulkan has to be specified in a bytecode
+format as opposed to human-readable syntax like
+https://en.wikipedia.org/wiki/OpenGL_Shading_Language[GLSL]
+https://shader-slang.org/slang/user-guide/[SLANG], and
+https://en.wikipedia.org/wiki/High-Level_Shading_Language[HLSL].
+This bytecode format is called https://www.khronos.org/spir[SPIR-V] and is designed
+to be used with both Vulkan and OpenCL (both Khronos APIs). It is a format that
```

Comment:
Maybe remove OpenCL. This is confusing esp. to beginners, as the SPIR generated for Vulkan is not directly compatible with the one for OpenCL.